python/pyarrow/pandas-shim.pxi

```diff
@@ -0,0 +1,172 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pandas lazy-loading API shim that reduces API call and import overhead
+
+cdef class _PandasAPIShim(object):
```

Comment:
I don't like it but get your point :)
We should have a ticket about refactoring this.

Reply:
Microperformance is a huge pain. The code that touches this is very sensitive to changes. If you _do_ refactor, you must be extremely paranoid about making things slower

For example I noticed that the operation

```
import pyarrow as pa
import numpy as np

arr = np.array([1, 2, 3, 4])[::2]
arrow_arr = pa.array(arr)
```

can get 10-20% slower with tiny modifications to things. Suppose that you have an operation that takes 3 microseconds, and you make changes that cause it to take 5 microseconds. Then you're dropping 2 seconds for every million calls. 

see https://mail.python.org/pipermail/pandas-dev/2016-March/000499.html